### PR TITLE
feat(context): introduce new binding scopes

### PR DESCRIPTION
**Pull request by raymondfeng**
Implements https://github.com/strongloop/loopback-next/issues/6325

# Current issues
Consider the following context hierarchy: invocationCtx -> requestCtx -> serverCtx -> appCtx, we use `Transient` scope for controllers/services so that each request will get a new instance. But if a controller/service is resolved within the `invocationCtx` (interceptors), a new instance will be created again. Ideally, we should get the same instance at the subtree of the `requestCtx`. Even worse, resolving a binding twice in the same `reqCtx` will get two different instances too.

Neither `Singleton` or `Context` can satisfy this requirement. Typically, controllers/servers are discovered and loaded into the application context. Those from components such as `RestComponent` also contribute bindings to the `appCtx` instead of `serverCtx`. `Singleton` will get one instance at the `appCtx` level while `Context` will get one instance per context.

# Proposed solution

When you build a chain of context objects, each context can define its `scope` property, such as `APPLICATION`, `SERVER`, and `REQUEST`. Then binding resolution will try to find the matching context with the same scope and use it as the key to share resolved values. This is probably the `CONTEXT` scope should have been.

Existing scopes such as `SINGLETON` and `TRANSIENT` will work as-is.

I'll update the docs if we agree on the direction.

# Questions

1. Do we want to make binding scopes extensible? At this point, it's an `enum`. We'll have change its type to `string` and introduce a `BindingScopeResolver`:

```
export interface BindingScopeResolver {
  getScopedContext(ctx: Context, scope: BindingScope): Context | undefined;
}
```

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: packages/rest/src/rest.server.ts [hunk 1]
```diff
@@ -211,6 +211,7 @@ export class RestServer
     config: RestServerConfig = {},
   ) {
     super(app);
+    this.scope = BindingScope.SERVER;
```
Comment (bajtos) [inline packages/rest/src/rest.server.ts:214]: I see we have to classes setting `scope = BindingScope.SERVER`:
- `RestServer`
- `ExpressServer`

In a typical REST application scaffolded via `lb4 app`, when I call `app.bind(...).inScope('BindingScope.SERVER')`, where will be the resolved value cached? At `RestServer` or `ExpressServer` level?

**Comment by bajtos** [comment]
The proposal looks reasonable :+1: I appreciate a detailed explanation in the pull request description 👏🏻 

> Do we want to make binding scopes extensible? At this point, it's an enum.

Personally, I am fine to use enum for now, until we have a requirement for a more flexible scope selector.

File: packages/rest/src/rest.server.ts (hunk 1 above)
Comment (raymondfeng) [reply packages/rest/src/rest.server.ts:214]: It depends on where the binding is resolved. 

- If the binding is resolved from the RestServer or below, it will be cached using the `RestServerContext` as the key
- If the binding is resolved from the ExpressServer or below, it will be cached using the `ExpressServerContext` as the key

On the same application, it's possible that both RestServer and ExpressServer have a shared/cached value for the same binding.

**Comment by raymondfeng** [comment]
@bajtos Another interesting question:

1. We would prefer to have controller bindings to use `Request` scope within a Rest application.
2. What if there is no `Request` context exists in the chain? For example, a lot of unit tests use `Application` only.
3. Should we report an error or fallback to `Transient` or `Context` scope? (Maybe `Context` makes sense)?

**Comment by mschnee** [comment]
I feel there could be some advantages to making binding scopes configurable, especially if you have designed a custom server implementatio; though I could also argue the converse point that in the case of a custom server/sequence, you should be creating and using a request-scope context.

**Comment by raymondfeng** [comment]
> I feel there could be some advantages to making binding scopes configurable, especially if you have designed a custom server implementatio; though I could also argue the converse point that in the case of a custom server/sequence, you should be creating and using a request-scope context.

There are two factors to allow custom binding scopes:

1. Relax the TypeScript typing for `BindingScope` - it's an `enum` at the moment.
2. Add a custom scope resolver. The following code in `Context` can be overridden (getResolutionContext is the key enabler to locate the correct context for binding/scope resolution).

```ts
  /**
   * Get the context matching the scope
   * @param scope - Binding scope
   */
  getScopedContext(
    scope:
      | BindingScope.APPLICATION
      | BindingScope.SERVER
      | BindingScope.REQUEST,
  ): Context | undefined {
    if (this.scope === scope) return this;
    if (this._parent) {
      return this._parent.getScopedContext(scope);
    }
    return undefined;
  }

  /**
   * Locate the resolution context for the given binding. Only bindings in the
   * resolution context and its ancestors are visible as dependencies to resolve
   * the given binding
   * @param binding - Binding object
   */
  getResolutionContext(
    binding: Readonly<Binding<unknown>>,
  ): Context | undefined {
    switch (binding.scope) {
      case BindingScope.SINGLETON:
        // Use the owner context
        return this.getOwnerContext(binding.key);
      case BindingScope.TRANSIENT:
      case BindingScope.CONTEXT:
        // Use the current context
        return this;
      default:
        // Use the scoped context
        return this.getScopedContext(binding.scope);
    }
  }
```

**Comment by bajtos** [comment]
> @bajtos Another interesting question:
> 
> 1. We would prefer to have controller bindings to use `Request` scope within a Rest application.
> 2. What if there is no `Request` context exists in the chain? For example, a lot of unit tests use `Application` only.
> 3. Should we report an error or fallback to `Transient` or `Context` scope? (Maybe `Context` makes sense)?

What if a controller depends on the current request object (receives it via dependency injection)? Such controller cannot be instantiated directly from `Application`. 

To enable unit tests, I think `RestApplication` (and/or `RestServer`) should provide API for creating Request-level context for the given a stub Request/Response. 

We can use `stubExpressContext` from `@loopback/testlab` to simulate an incoming HTTP request. https://github.com/strongloop/loopback-next/blob/95b6a2b7ce64e614720df43b905f77a53a54e438/packages/testlab/src/shot.ts#L95-L97

Would something like this 👇🏻 work?

```ts
// implementation
class RestServer {
  createStubRequestContext(requestOptions: ShotRequestOptions) {
    // create request context
    // setup per-request bindings
    // return the context to the caller
  }
}

// in the tests
const ctx = restApp.restServer.createStubRequestContext({url: '/products/1'});
const productController = await ctx.get<ProductController>('controllers.ProductController');
// etc.
```


We can draw inspiration from Fastify's `inject` API, see https://www.fastify.io/docs/latest/Testing/#benefits-of-using-fastifyinject

---

Setting the unit-testing scenario aside, you discovered an edge case I was thinking about for past few days too. It would be great if we can report a descriptive error when the user attempts to resolve a value from application-level context when the binding is in per-request scope. For example something along the following lines:

> Cannot resolve "controllers.TodoController" from context scope "Application". The binding has scope "Request".

**Comment by raymondfeng** [comment]
> What if a controller depends on the current request object (receives it via dependency injection)? Such controller cannot be instantiated directly from Application.

We have two related things here:

1. Decide which context should be used for resolution

When LB tries to resolve a controller binding key using a context object, we first use the scope to find out which context in the chain should be the starting point for searching (only itself and its ancestors).

My question is whether we should fall back to the current context if no context in the chain matches the `REQUEST` scope.

There are possibly three contexts involved:

- The current context (used to initialize the resolution)
- The resolution context (used to scope the resolution visibility)
- The owner context (where the binding is contained)

The resolution context is determined as follows:

- The current context for `CONTEXT` and `TRANSIENT` scopes 
- The owner context for `SINGLETON` scope
- The matching context for other scopes

2. Inject a dependency into a controller

- For constructor or property injection, we'll use the resolution context of the controller binding as the current context.
- For method parameter injections, the invocation context will be used as the current context.

In both cases, we'll report errors if a binding key cannot be resolved in the resolution context or its ancestors. For example, a singleton binding registered at `application` level cannot have dependencies to be injected from `server` level.  

If the dependency cannot be resolved within the

File: packages/rest/src/rest.server.ts (hunk 1 above)
Comment (bajtos) [reply packages/rest/src/rest.server.ts:214]: I thought that `RestServer` is using `ExpressServer` under the hood, that's why I was curious about where a value will be cached. I checked the sources and see that `ExpressServer` is not used by `RestServer` (or anywhere else in the typical LB4 stack), in which case my question is not relevant.

It does make me wonder though why are we providing & maintaining `ExpressServer` when we are not using it ourselves 🤷🏻

File: docs/site/Binding.md [hunk 2]
```diff
@@ -379,14 +382,154 @@ To understand the difference between `@bind()/@injectable` and `ctx.bind()`, see
...
-example, a logging provider can be refreshed to pick up a new logging level. The
-same functionality can be achieved with `TRANSIENT` scope but with much more
-overhead.
+```
+invocationCtx -> requestCtx -> serverCtx -> appCtx
+```
+
+We use `TRANSIENT` scope for controllers/services so that each request will get
+a new instance. But if a controller/service is resolved within the
+`invocationCtx` (by interceptors), a new instance will be created again.
```
Comment (jannyHou) [inline docs/site/Binding.md:395]: > a new instance will be created again.

Is it because each interceptor will resolve a controller/service instance?

File: docs/site/Binding.md [hunk 3]
```diff
@@ -379,14 +382,154 @@ To understand the difference between `@bind()/@injectable` and `ctx.bind()`, see
...
+controllers/servers are discovered and loaded into the application context.
+Those from components such as RestComponent also contribute bindings to the
+`appCtx` instead of `serverCtx`. With `SINGLETON` scope, we will get one
+instance at the `appCtx` level. With `CONTEXT` scope, we will get one instance
+per context. A set of fine-grained scopes has been introduced to allow better
+scoping of binding resolutions.
+
+- BindingScope.APPLICATION
+- BindingScope.SERVER
+- BindingScope.REQUEST
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply docs/site/Binding.md:410]: `Request` is the mostly used by other IoC frameworks too. We can also interpret an event handling as `event request processing`. Let's stick with `REQUEST` for now unless somebody can come up a better one.

File: docs/site/Binding.md (hunk 2 above)
Comment (raymondfeng) [reply docs/site/Binding.md:395]: I improved the docs with an example to make it even more clear.

**Comment by raymondfeng** [comment]
@jannyHou @bajtos PTAL.

File: packages/context/src/binding.ts [hunk 4]
```diff
@@ -104,6 +109,62 @@ export enum BindingScope {
    * - req2.get('b1') ==> 0 (always)
    */
   SINGLETON = 'Singleton',
+
+  /*
```
Comment (jannyHou) [inline packages/context/src/binding.ts:113]: Had a chat with @raymondfeng regarding the new scopes. We probably want to mention that, when the specified scope is same as owner, it's equivalent as scope `SINGLETON`.

E.g.

```ts
    appCtx
      .bind('foo')
      .toDynamicValue(() => `foo.app.${generateUniqueId()}`)
      .inScope(BindingScope.APPLICATION);
```
is equivalent to 
```ts
    appCtx
      .bind('foo')
      .toDynamicValue(() => `foo.app.${generateUniqueId()}`)
      .inScope(BindingScope.SINGLETON);
```

**Review by jannyHou — APPROVED** [review]
👍 Great improvement! LGTM.

**Comment by bradharms** [comment]
This deprecates the CONTEXT scope, which our project is depending on for a custom authentication system we've been implementing in our project. (See [this stack overflow question](https://stackoverflow.com/questions/64053316/prototypical-inheritance-and-extension-in-the-configuration-for-the-service-mana))

TL;DR, we wanted to use a sub-context of the current request context as a means of generating alternate versions of existing services. The only difference between the original service and the alternate version is that the dependencies they get injected with are provided by different providers than in the root context.

**Comment by raymondfeng** [comment]
If the sub-context of the current request context is going to be the leaf context, you can use `SINGLETON` scope to bind to the sub-context.

**Comment by bradharms** [comment]
In practice it turns out part of the context is resolved before the request context and part within the request context. This is because some of the services are middleware which run at the sequence level, and others are injected into a controller.

**Comment by raymondfeng** [comment]
@bradharms Please be aware that middleware can be bound in `Request` or `Transient` scope so that it will receive the request context via dependency injection too. 

If you can elaborate your use case, we can settle on a reasonable solution and decide if `Context` scope should be deprecated or not.

**Comment by bradharms** [comment]
My apologies. The middleware I'm referring to are not the rest middleware that were introduced in @loopback/rest 6.0. They are a home grown version of middleware we created prior to the introduction of Loopback's own version of middleware and they are executed in an action that is part of a custom sequence. We are still in the process of converting our middleware to use the standard loopback style so that we can drop the custom middleware and use the standard loopback middleware-based sequence. Once this is completed we will re-evaluate.

**Comment by raymondfeng** [comment]
@bradharms That makes sense now. In fact, the binding scopes in LB4 now is close to be extensible with this PR - we just have to extend the `BindingScope` enum if needed. Each context can now tell the resolution if it matches a given scope.

**Comment by bradharms** [comment]
I don't understand what you mean by extending BindingScope. Wouldn't it be more flexible to simply allow consumers to specify an arbitrary string key to identify the scope so they can add their own scopes for specific contexts?

**Comment by raymondfeng** [comment]
We decided not to relax `BindingScope` to `string` for now unless we have compelling use cases to do so.
